fix(hash): recreate container on project config content change

idsulik opened the pull request:
**What I did**
Fixed `hash.ServiceHash()` to support config content change

**Related issue**
https://github.com/docker/compose/issues/11900

![image](https://github.com/docker/compose/assets/3595194/4044449c-e6f8-4463-81e3-81e98980947f)

ndeloof commented:
While I understand the intent, I don't like we get the config content added into the service hash. This also only makes sense as the config content is inlined. 
I wonder we could rely on a label to track the config state by the time it was created : `com.docker.compose.config.name=<config hash>`.

ndeloof commented on pkg/compose/hash.go:
```diff
@@ -37,5 +37,12 @@ func ServiceHash(o types.ServiceConfig) (string, error) {
+
+	for _, serviceConfig := range o.Configs {
+		if projectConfig, ok := project.Configs[serviceConfig.Source]; ok {
+			bytes = append(bytes, []byte(projectConfig.Content)...)
```
This won't detect changes for config set by environment of file

idsulik replied:
added support for env and config file

idsulik commented:
> I don't like we get the config content added into the service hash

why? 



> I wonder we could rely on a label to track the config state by the time it was created : `com.docker.compose.config.name=<config hash>`.

I don't get the idea, _time it was created_ - do you mean config file? or docker-compose.yaml? but docker-compose can refer to external config file

ndeloof commented:
time _container_ was created, so we can check it needs to be recreated if current config doesn't match

jhrotko commented on pkg/compose/hash.go:
```diff
@@ -37,5 +38,22 @@ func ServiceHash(o types.ServiceConfig) (string, error) {
 	}
+
+	for _, serviceConfig := range o.Configs {
+		if projectConfig, ok := project.Configs[serviceConfig.Source]; ok {
```
nit: I would invert the if statements so that we do not get the code so nested

idsulik replied:
@jhrotko done

ndeloof commented:
Some poins:
1. At some point we will have to do the same for secrets (same constraints) and networks/volumes (which can get labeled with a has). Solution we adopt here should be extensible to allow this in the future
2. Configs set by file are implemented by a bind mount but this may change in near future (see https://github.com/docker/compose/pull/11871) so we also need to consider those
3. There's no technical reason docker engine can't offer secrets/configs natively, just this is guarded by Swarm mode, but AFAIK some discussion happened to get them _also_ available in standalone mode. I can't tell if/when this would take place, but preferably the logic here should consider this may happen
4. Last but not least, I'd prefer we don't mix service config hash with resources it depends on, so my suggestion to introduce an additional label `com.docker.compose.config.xx=hash` that we can use to track this relation, and need to recreate container, without the need to change the service hash computation (which as impact on existing installations)

idsulik commented:
@ndeloof thanks for the details. pushed changes:
1. reverted old changes
2. added new func `ServiceDependenciesHash` and label 	
```
// ConfigHashDependenciesLabel stores configuration hash for a compose service dependencies
ConfigHashDependenciesLabel = "com.docker.compose.config-hash-dependencies"
```

ndeloof commented on pkg/compose/hash.go:
```diff
@@ -39,3 +40,28 @@ func ServiceHash(o types.ServiceConfig) (string, error) {
+		if projectConfig.Content != "" {
+			bytes = append(bytes, []byte(projectConfig.Content)...)
+		} else if projectConfig.File != "" {
+			content, err := os.ReadFile(projectConfig.File)
```
`File` could also be a folder. Maybe an option would be to read content into a tar (code exists to inject secrets) and compute hash from it

While the general approach looks good, I'd prefer we have a more explicit hash and label _per resource_, and a dedicated `hash` function for configs (and secrets), so that it make it easier to diagnose use of those labels

idsulik commented:
Up)

idsulik commented:
Hi @ndeloof  ,

I noticed that this pull request has been pending for a while. Is there anything more needed from my side? Please let me know if there are any updates or additional changes required.

Thanks!

idsulik commented:
Hi @ndeloof ,

I noticed that this pull request has been pending for a while. Is there anything more needed from my side? Please let me know if there are any updates or additional changes required.

Thanks!

idsulik commented:
Hi @ndeloof ,

I noticed that this pull request has been pending for a while. Is there anything more needed from my side? Please let me know if there are any updates or additional changes required.

Thanks!

idsulik commented:
Hi @ndeloof ,

I noticed that this pull request has been pending for a while. Is there anything more needed from my side? Please let me know if there are any updates or additional changes required.

Thanks!

idsulik commented:
Hi @ndeloof , what do you think about think about the PR?

ndeloof commented on pkg/compose/hash.go:
```diff
@@ -40,3 +46,122 @@ func ServiceHash(o types.ServiceConfig) (string, error) {
+				return err
+			}
+			if !info.IsDir() {
+				fileContent, err := os.ReadFile(path)
```
this will not detect file being rename. Maybe better to create a tar stream from file/dir content

idsulik replied:
@ndeloof it creates a tar from the content here pkg/compose/hash.go:122. 

`b, err := utils.CreateTar(content, serviceConfig, modTime)`

idsulik replied:
@ndeloof is there any reason to create tar inside a tar?   the `createTarForConfig` func calls this `readPathContent` func and creates a tar with the path content

ndeloof replied:
here you are creating a tar with all content appended as []byte, this doesn't reflect the filesystem structure for intput. You have to create one tar entry per file with file/folder name.

idsulik replied:
@ndeloof you're right, thank you! Updated the code, now it creates a tar file for a given path and keeps file structure

ndeloof commented on pkg/compose/hash.go:
```diff
@@ -34,9 +38,75 @@ func ServiceHash(o types.ServiceConfig) (string, error) {
+		data = append(data, b.Bytes()...)
+	}
+
+	return digest.SHA256.FromBytes(data).Encoded(), nil
```
I'd prefer we have one label per config/secret mount, so it makes it easier to track|debug changes and container being recreated.
Also need to consider config can be mounted from docker host, i.e. file is not available for compose to compute hash, and then must be excluded from label / no label created

idsulik replied:
do you mean something like that:
```
com.docker.compose.service.configs-hash-{configName}={hash}
com.docker.compose.service.configs-hash-{serviceName}={hash}
```
?

ndeloof replied:
indeed, or maybe, to follow the dot-notation style used for labels, `com.docker.compose.service.configs.{configName}.hash`

idsulik replied:
It seems to me that this will complicate the logic, first you need to generate a hash for each item separately, then you need to go through all labels whose names start with “com.docker.compose.service.configs.” to check if the hash has changed.

ndeloof replied:
Doesn't look such a pain to me, as this would allow to trace reason we recreate a container, and make it easier to diagnose potential regressions (this sometimes happened :P)

```go
for c := range service.Configs {
  hash := labels["com.docker.compose.configs."+c+".hash
  expected := ConfigHash(project.Configs[c]
  if hash := expected {
    log.Debug("container has to be recreated after config %s has been updated", c)
    return DIVERGED
  }
}
```

idsulik replied:
@ndeloof, but we don't have a config/service name that we can use to create the label name. I pushed changes to create a hash of the config/secret of each service so it'll be easy to figure out what service's config/secret caused the change

apollo13 commented:
This would be a really great addition, anything I can help out with?

idsulik commented:
> This would be a really great addition, anything I can help out with?

no, thank you, waiting for @ndeloof response

apollo13 commented:
@idsulik Maybe it might make sense to fix the failing linting checks before another round of reviews?

idsulik commented:
@ndeloof could you please check it?

idsulik commented:
@glours hi! Maybe you can check it? otherwise, it can be stuck here for a long time

idsulik commented:
@ndeloof @glours hi! Rebased the main branch and resolved conflicts, is there anything else I should do ?

idsulik commented:
I have to resolve conflicts several times, could you please check it? @ndeloof
